src/Microsoft.Tye.Hosting/IngressService.cs

```diff
@@ -0,0 +1,287 @@
+
+                    builder.Logging.AddProvider(new ServiceLoggerProvider(service.Logs));
+
+                    var addresses = new List<string>();
+
+                    // Bind to the addresses on this resource
+                    for (int i = 0; i < serviceDescription.Replicas; i++)
+                    {
+                        // Fake replicas since it's all running processes
+                        var replica = service.Description.Name + "_" + Guid.NewGuid().ToString().Substring(0, 10).ToLower();
+                        var status = new IngressStatus(service, replica);
+                        service.Replicas[replica] = status;
+
+                        var ports = new List<int>();
+
+                        foreach (var binding in serviceDescription.Bindings)
+                        {
+                            if (binding.Port == null)
+                            {
+                                continue;
+                            }
+
+                            var port = service.PortMap[binding.Port.Value][i];
+                            ports.Add(port);
+                            var url = $"{binding.Protocol ?? "http"}://localhost:{port}";
+                            addresses.Add(url);
+                        }
+
+                        status.Ports = ports;
+
+                        service.ReplicaEvents.OnNext(new ReplicaEvent(ReplicaState.Added, status));
+                    }
+
+                    builder.Server.UseUrls(addresses.ToArray());
+                    var webApp = builder.Build();
+
+                    _webApplications.Add(webApp);
+
+                    // For each ingress rule, bind to the path and host
+                    foreach (var rule in runInfo.Rules)
+                    {
+                        if (!application.Services.TryGetValue(rule.Service, out var target))
+                        {
+                            continue;
+                        }
+
+                        _logger.LogInformation("Processing ingress rule: Path:{Path}, Host:{Host}, Service:{Service}", rule.Path, rule.Host, rule.Service);
+
+                        var targetServiceDescription = target.Description;
+
+                        var uris = new List<Uri>();
+
+                        // For each of the target service replicas, get the base URL
+                        // based on the replica port
+                        for (int i = 0; i < targetServiceDescription.Replicas; i++)
+                        {
+                            foreach (var binding in targetServiceDescription.Bindings)
+                            {
+                                if (binding.Port == null)
+                                {
+                                    continue;
+                                }
+
+                                var port = target.PortMap[binding.Port.Value][i];
+                                var url = $"{binding.Protocol ?? "http"}://localhost:{port}";
```

Comment:
I see this pattern of `binding.Protocol ?? "http"` everywhere. Can we just default to http if it isn't set?

Reply:
If we ever need to infer anything from "not set" vs "explicitly set to http" then it's potentially an issue.